src/nvim/log.c

```diff
@@ -124,10 +128,27 @@ void log_unlock(void)
 /// @param line_num   Source line number, or -1
 /// @param eol        Append linefeed "\n"
 /// @param fmt        printf-style format string
+///
+/// @return true if log was emitted normally, false if failed or recursive
 bool logmsg(int log_level, const char *context, const char *func_name, int line_num, bool eol,
             const char *fmt, ...)
   FUNC_ATTR_UNUSED FUNC_ATTR_PRINTF(6, 7)
 {
+  static int recursive = 0;
+  if (!did_log_init) {
+    g_stats.log_skip++;
+    // set_init_1 may try logging before we are ready. 6f27f5ef91b3 #10183
+    return false;
+  }
+  if (recursive) {
+    recursive++;
+    if (recursive <= 2) {
+      siemsg("E5430: %s:%d: recursive log!", func_name, line_num);
```

Comment:
show a user error (once) instead of logging to stderr (which is unpleasant for embedders?)